File: LeetCodeProblemsGo/LeetCode1115PrintFooBarAlternately/LeetCode1115PrintFooBarAlternately.go
package LeetCode1115PrintFooBarAlternately

import "sync"

type FooBar struct {
	n        int
	fooMutex sync.Mutex
	barMutex sync.Mutex
}

func NewFooBar(n int) *FooBar {
	result := &FooBar{n: n}
	result.barMutex.Lock()
	return result
}

func (fb *FooBar) Foo(printFoo func()) {
	for i := 0; i < fb.n; i++ {
		fb.fooMutex.Lock()
		// printFoo() outputs "foo". Do not change or remove this line.
		printFoo()
		fb.barMutex.Unlock()
	}
}

func (fb *FooBar) Bar(printBar func()) {
	for i := 0; i < fb.n; i++ {
		fb.barMutex.Lock()
		// printBar() outputs "bar". Do not change or remove this line.
		printBar()
		fb.fooMutex.Unlock()
	}
}

File: LeetCodeProblems/Problems/LeetCode1115PrintFooBarAlternately.cs
namespace LeetCodeProblems.Problems;
using System.Threading;
public class LeetCode1115PrintFooBarAlternately
{
    public class FooBar {
        private int n;
        SemaphoreSlim fooSemaphore = new SemaphoreSlim(1, 1);
        SemaphoreSlim barSemaphore = new SemaphoreSlim(0, 1);

        public FooBar(int n) {
            this.n = n;
        }

        public void Foo(Action printFoo) {
        
            for (int i = 0; i < n; i++) {
            
                // printFoo() outputs "foo". Do not change or remove this line.
                fooSemaphore.Wait();
                printFoo();
                barSemaphore.Release();
       
            }
        }

        public void Bar(Action printBar) {
        
            for (int i = 0; i < n; i++) {
            
                // printBar() outputs "bar". Do not change or remove this line.
                barSemaphore.Wait();
                printBar();
                fooSemaphore.Release();
            }
        }
    }
}